File: gogit/git/commit.go
package git

import "bytes"
import "fmt"

type Commit struct {
	oid     []byte
	tree    []byte
	author  Author
	message string
}

func (commit *Commit) Init(tree []byte, author Author, msg string) *Commit {
	commit.tree = tree
	commit.author = author
	commit.message = msg
	return commit
}

func (commit Commit) Type() string {
	return "commit"
}

func (commit Commit) Data() []byte {
	var buf bytes.Buffer
	buf.WriteString(fmt.Sprintf("tree %x\n", commit.tree))
	buf.WriteString(fmt.Sprintf("author %s\n", commit.author.String()))
	buf.WriteString(fmt.Sprintf("committer %s\n", commit.author.String()))
	buf.WriteString("\n")
	buf.WriteString(commit.message)
	return buf.Bytes()
}

func (commit Commit) Id() []byte {
	return commit.oid
}

func (commit *Commit) SetId(oid []byte) {
	commit.oid = oid
}

File: cppgit/commit.cc
#include "./commit.hh"

namespace git {

const_buffer Commit::data() const {
  using std::endl;

  stringstream stream;

  stream << "tree " << tree_ << endl;
  stream << "author " << author_.str() << endl;
  stream << "committer " << author_.str() << endl;
  stream << endl;
  stream << message_;
  stream << endl;

  string data{stream.str()};
  char* buffer = new char[data.size()];
  data.copy(buffer, data.size());
  return {buffer, data.size()};
}

}  // namespace git
